### PR TITLE
Feature/tree component

### PR DESCRIPTION
**Pull request by llali**
Added tree component to model view builder .
The api to create the tree is using the vs code TreeDataProvider but the new tree component also support having a checkbox in the node. Here's a sample of how to create the tree component:

`
let root = TreeNode.createTree(treeData);

		let treeDataProvider = new TreeDataProvider(root);

		let tree: sqlops.TreeComponent<TreeNode> = view.modelBuilder.tree<TreeNode>().withProperties({
			'withCheckbox': true
		}).component();
		tree.registerDataProvider(treeDataProvider);
`

The tree's checkbox support intermediate state which means if some of the children of a node are checked, the parent's checkbox shows an intermediate state. Also checking/unchecking a parent make all children checked/checked

image

### CODE REVIEW

**Comment by llali** [comment]
@MattIrv I'm adding a new option to specify if the tree has checkbox or not.
for using the existing vs code API I'm not sure if I can do it because I want to add the tree to dialog and wizard and I don't think I can do that with vs code APIs. is there any other way I can benefit from those APIs?

**Comment by MattIrv** [comment]
@llali would it be possible to extend those interfaces with whatever checkbox-specific properties we need and then use `TreeDataProvider` (or our extended version of it) as the input for the component instead of our own `TreeComponentDataModel`?

File: src/sql/parts/modelComponents/tree/tree.component.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+			this.createTreeControl();
+
+		}
+	}
+
+	ngOnDestroy(): void {
+		this.baseDestroy();
+	}
+
+	public setDataProvider(context: any): any {
```
[the remark this answers is not in the document]
Comment (llali) [reply src/sql/parts/modelComponents/tree/tree.component.ts]: the type if different for each data provider

---
In reply to: [207068145](https://github.com/Microsoft/sqlopsstudio/pull/2077#discussion_r207068145) [](ancestors = 207068145)

File: src/sql/parts/modelComponents/tree/treeComponentRenderer.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+			this._onChange.fire(this._checkbox.checked);
+			if (this.model && this.model.onCheckedChanged) {
+				this.model.onCheckedChanged(this._checkbox.checked);
+			}
+		});
+	}
+
+	public get checkboxState(): TreeCheckboxState {
+		if (this._checkbox.indeterminate) {
+			return TreeCheckboxState.Intermediate;
```
[the remark this answers is not in the document]
Comment (llali) [reply src/sql/parts/modelComponents/tree/treeComponentRenderer.ts]: they are technically the same thing

---
In reply to: [207068924](https://github.com/Microsoft/sqlopsstudio/pull/2077#discussion_r207068924) [](ancestors = 207068924)

File: src/sql/parts/modelComponents/tree/treeDataSource.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+				return this._dataProvider.getChildren(undefined);
+			} else {
+				return this._dataProvider.getChildren(node);
+			}
+		}
+		return TPromise.as([]);
+	}
+
+	public getParent(tree: ITree, node: any): TPromise<any> {
+		return TPromise.as(null);
```
[the remark this answers is not in the document]
Comment (llali) [reply src/sql/parts/modelComponents/tree/treeDataSource.ts]: doesn't need to do anything with parent right now 

---
In reply to: [207069396](https://github.com/Microsoft/sqlopsstudio/pull/2077#discussion_r207069396) [](ancestors = 207069396)

File: src/sql/workbench/api/node/extHostModelViewTree.ts [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+	constructor(
+		private _mainContext: IMainContext
+	) {
+		this._proxy = this._mainContext.getProxy(SqlMainContext.MainThreadModelView);
+	}
+
+	$createTreeView<T>(handle: number, componentId: string, options: { treeDataProvider: sqlops.TreeComponentDataProvider<T> }): vscode.TreeView<T> {
+		if (!options || !options.treeDataProvider) {
+			throw new Error('Options with treeDataProvider is mandatory');
```
[the remark this answers is not in the document]
Comment (llali) [reply src/sql/workbench/api/node/extHostModelViewTree.ts]: user won't see this

---
In reply to: [207069746](https://github.com/Microsoft/sqlopsstudio/pull/2077#discussion_r207069746) [](ancestors = 207069746)

**Pull request merged**